google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/apis/CloudApiManager.java

```diff
@@ -69,67 +75,99 @@ static void enableApis(Set<CloudLibrary> libraries, CloudProject cloudProject, P
       return;
     }
 
-    ServiceManagement serviceManagement =
-        GoogleApiClientFactory.getInstance().getServiceManagementClient(user.get().getCredential());
-
-    List<CloudLibrary> enabledApis = Lists.newArrayList();
-    List<String> erroredApiNames = Lists.newArrayList();
-
-    libraries.forEach(
-        library -> {
-          try {
-            serviceManagement
-                .services()
-                .enable(
-                    library.getServiceName(),
-                    new EnableServiceRequest()
-                        .setConsumerId(
-                            String.format(
-                                SERVICE_REQUEST_PROJECT_PATTERN, cloudProject.projectId())))
-                .execute();
-            enabledApis.add(library);
-          } catch (IOException e) {
-            LOG.warn(
-                "Exception occurred attempting to enable API " + library.getName() + " on GCP", e);
-            erroredApiNames.add(library.getName());
-          }
-        });
-
-    if (!erroredApiNames.isEmpty()) {
-      notifyApiEnableError(erroredApiNames, project);
+    List<CloudLibrary> libraryList = new ArrayList<>(libraries);
+    Set<CloudLibrary> enabledApis = Sets.newHashSet();
+    Set<CloudLibrary> erroredApis = Sets.newHashSet();
+
+    for (int i = 0; i < libraryList.size(); i++) {
+      CloudLibrary library = libraryList.get(i);
+
+      try {
+        ProgressIndicator progress = ProgressManager.getInstance().getProgressIndicator();
+
+        if (progress.isCanceled()) {
+          LOG.info("API enablement canceled by user");
+          notifyApiEnableSkipped(Sets.difference(libraries, enabledApis), project);
+          return;
+        }
+
+        setApiEnableProgress(
+            progress,
+            library.getName(),
+            cloudProject.projectName(),
+            (double) i / libraryList.size());
+        enableApi(library, cloudProject, user.get());
```

Comment:
What could be an average expected time to enable APIs on GCP? It looks like we do this synchronously here, it's probably fine if this is quick, but may be annoying if it takes longer than 10-20 seconds for example. Could we also look at `runProcessWithProgressAsynchronously` in `ProgressManager`?

Reply:
from my tests. each api takes about a second or less to enable

I could explore doing this asynchronously in a separate PR though since that may be a better model for this regardless.

Reply:
sounds good, just to be sure if some user runs into issue and we freeze his IDE for a few minutes

Reply:
+1. There is always the cancel button escape hatch though, just to point that out. I still agree though.